library/src/main/java/com/esafirm/rxdownloader/RxDownloader.java

```diff
@@ -47,44 +48,108 @@ private DownloadManager getDownloadManager() {
 
-    public Observable<String> download(@NonNull String url,
-                                       @NonNull String filename,
-                                       @NonNull String mimeType,
-                                       boolean showCompletedNotification) {
+    public Single<String> download(@NonNull String url,
+                                             @NonNull String filename,
+                                             @NonNull String mimeType,
+                                             boolean showCompletedNotification) {
         return download(createRequest(url, filename, null,
                 mimeType, true, showCompletedNotification));
     }
 
-    public Observable<String> download(@NonNull String url,
-                                       @NonNull String filename,
-                                       @NonNull String destinationPath,
-                                       @NonNull String mimeType,
-                                       boolean showCompletedNotification) {
+    public Single<String> download(@NonNull String url,
+                                             @NonNull String filename,
+                                             @NonNull String destinationPath,
+                                             @NonNull String mimeType,
+                                             boolean showCompletedNotification) {
         return download(createRequest(url, filename, destinationPath,
                 mimeType, true, showCompletedNotification));
     }
 
-    public Observable<String> downloadInFilesDir(@NonNull String url,
-                                                 @NonNull String filename,
-                                                 @NonNull String destinationPath,
-                                                 @NonNull String mimeType,
-                                                 boolean showCompletedNotification) {
+    public Single<String> downloadInFilesDir(@NonNull String url,
+                                                       @NonNull String filename,
+                                                       @NonNull String destinationPath,
+                                                       @NonNull String mimeType,
+                                                       boolean showCompletedNotification) {
         return download(createRequest(url, filename, destinationPath,
                 mimeType, false, showCompletedNotification));
     }
 
-    public Observable<String> download(DownloadManager.Request request) {
+    public Single<String> download(DownloadManager.Request request) {
         long downloadId = getDownloadManager().enqueue(request);
 
-        PublishSubject<String> publishSubject = PublishSubject.create();
-        subjectMap.put(downloadId, publishSubject);
+        PublishSubject<DownloadState> publishSubject = PublishSubject.create();
+        progressSubjectMap.put(downloadId, publishSubject);
+
+        return publishSubject
```

Comment:
We could have one `Map` of `Subject<DownloadState>`s   
What i mean by the original approach is that we don't need to include the progress checking in simple `download` method

Have you take a look at this sample -> https://github.com/esafirm/kotlin-playground/blob/master/src/main/kotlin/nolambda.playground/rx/RxDownloaderTick.kt

We could have two `Observable` for two different cases 
1. Listen to progress (+ if progress is max it also considered as complete) 
2. Listen to complete event 

So for simple `download` we could only rely to the "Listen to complete event" `Observable` 

And for the `downloadWithProgress` while it could work with only the first, we should add the "listen to complete event" to for the reason that i mention earlier

Reply:
Ah, gotchya. I think I get it, though I'll need a play about. I'm not sure when I'm able to give this attention, to be honest due to busy time at work due to deadlines, just to let you know.